### PR TITLE
align libai linear with nn.Linear

### PR DESCRIPTION
**Pull request by L1aoXingyu**

### CODE REVIEW

**Comment by rentainhe** [comment]
星宇可以顺便帮忙把这两行删了嘛2333我多加了2行在swin的config里
https://github.com/Oneflow-Inc/libai/blob/d4a6803637f7662ffb30c321e318e05363c3a37e/configs/swin_imagenet.py#L15

**Comment by L1aoXingyu** [comment]
> 星宇可以顺便帮忙把这两行删了嘛2333我多加了2行在swin的config里
> 
> https://github.com/Oneflow-Inc/libai/blob/d4a6803637f7662ffb30c321e318e05363c3a37e/configs/swin_imagenet.py#L15

好的

File: libai/layers/linear.py [hunk 1]
```diff
@@ -22,16 +22,16 @@
 
 class Linear1D(nn.Module):
     r"""Linear layer with 1D parallelism which includes column parallelism and row parallelism.
-    The linear layer is defined as :math:`Y = XA + b`.
+    The linear layer is defined as :math:`Y = xA^T + b`.
```
Comment (Ldpe2G) [inline libai/layers/linear.py:25]: 大写 'X' ？

File: tests/layers/test_linear.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+from oneflow import nn
+
+from libai.layers import Linear
+from libai.utils import distributed as dist
+
+
+class TestLinear(flow.unittest.TestCase):
+    def setUp(self):
+        # input with shape (8, 8)
+        self.inputs = flow.tensor(
```
Comment (Ldpe2G) [inline tests/layers/test_linear.py:32]: 这里输入和权值不采用随机的原因是？

File: libai/layers/linear.py [hunk 3]
```diff
@@ -72,12 +72,18 @@ def __init__(
...
+            # so weight sbp sign actually be [B, S(1)]
+            # bias sbp sign: [B, S(0)]
+            weight_sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.split(0)])
             bias_sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.split(0)])
         elif parallel == "row":
-            # Row parallel weight sbp: [B, S(0)] and bias sbp: [B, B]
-            weight_sbp = dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.split(0)])
+            # Row parallel
+            # weight sbp sign: [B, S(1)], weight will be transposed when performing matmul
+            # so weight sbp sign actually be [B, S(1)]
```
[the remark this answers is not in the document]
Comment (L1aoXingyu) [reply libai/layers/linear.py:84]: 这里其实描述的是 W^T 的 sbp，因为现在的计算公式和 nn.Linear 对齐，是 xW^T + b

[the remark this answers is not in the document]
Comment (rentainhe) [reply libai/layers/linear.py:84]: OK~

File: libai/layers/linear.py (hunk 1 above)
Comment (L1aoXingyu) [reply libai/layers/linear.py:25]: image

这里对齐 nn.Linear 的文档，我把 y 也改成小写

File: tests/layers/test_linear.py (hunk 2 above)
Comment (L1aoXingyu) [reply tests/layers/test_linear.py:32]: 应该是当时写两卡模型并行的 linear 时候，发现和单卡对齐有点麻烦，应该可以使用随机的输入和 weight，我尝试改一下

Comment (L1aoXingyu) [reply tests/layers/test_linear.py]: 已经使用了随机值，可以再看看

**Review by Ldpe2G — APPROVED** [review]

**Review by rentainhe — APPROVED** [review]

**Comment by L1aoXingyu** [comment]
image

在 loss 对齐分支完成 bert 的验证
